packages/teams-js/src/internal/communication.ts

```diff
@@ -305,6 +312,70 @@ function processMessage(evt: DOMMessageEvent): void {
 
 const shouldProcessMessageLogger = communicationLogger.extend('shouldProcessMessage');
 
+function sendValidateDomainRequest(messageOrigin: URL): Promise<boolean> {
+  const messageOriginUrl = sanitizeAndStrigifyURL(messageOrigin);
+  return sendAndUnwrap('validateDomains', messageOriginUrl);
+}
+
+export function validateOrigin(messageOrigin: URL): boolean {
+  // Check whether the url is in the pre-known allowlist or supplied by user
+  if (!isValidHttpsURL(messageOrigin)) {
+    validateOriginLogger(
+      'Origin %s is invalid because it is not using https protocol. Protocol being used: %s',
+      messageOrigin,
+      messageOrigin.protocol,
+    );
+    return false;
+  }
+  const messageOriginHost = messageOrigin.host;
+  let isVerified = false;
+  if (Communication.alreadyVerifiedDomains.has(messageOriginHost)) {
+    isVerified = Communication.alreadyVerifiedDomains.get(messageOriginHost);
+  } else {
+    sendValidateDomainRequest(messageOrigin).then((hubResponse: boolean) => {
+      isVerified = hubResponse;
+      Communication.alreadyVerifiedDomains.set(messageOriginHost, hubResponse);
+    });
+  }
+  if (isVerified) {
+    return true;
+  }
+  if (fallbackValidateOrigin(messageOriginHost)) {
+    return true;
+  }
+  if (validateAdditionalOrigins(messageOriginHost)) {
+    return true;
+  }
```

Comment:
Is this working the way you expect? I would think that all the code after the else block (e.g., starting with `if (isVerified)` is going to run _before_ the code inside the `then` block. If you need to guarantee that code runs **after** calling an async/`Promise` it needs to be in the `then` block.

Try running this sample and check out the ordering of the console statements that are output:

[https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABDMYCmAnAFAZwBZwgA2AJgEpoBWa0AXIgEZxxFoCGYAlPQAoZwBbGDjQAeAG5wYJAHyIA3gFgAUIkQY0UEBiT5CpCtWiIA-Ij6DhaAHQajULJ0S9+QkbbQ4W4tI4DcKgC+KiqgkLAIiIRQmI4uliKiTCzsYHJKqsioscBsRCKc1lB4aGBYjogAvOmIEAherNZEcADmWABExaWMzRAA1u2cfoiBhRBsUBB45RpsXmD0HACeTtUKdWANNs1t7eOTeD1w-YN+oypqG1tNrR0AInCeiMXCiC2aiDvvJIgAQgCiADEAPJkf5RDCINjAGKQgDkXSQDF6fThplOF3Umm0SAsbhsGgaPiwUAwIDQQyCIWU0ViQyAA](https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABDMYCmAnAFAZwBZwgA2AJgEpoBWa0AXIgEZxxFoCGYAlPQAoZwBbGDjQAeAG5wYJAHyIA3gFgAUIkQY0UEBiT5CpCtWiIA-Ij6DhaAHQajULJ0S9+QkbbQ4W4tI4DcKgC+KiqgkLAIiIRQmI4uliKiTCzsYHJKqsioscBsRCKc1lB4aGBYjogAvOmIEAherNZEcADmWABExaWMzRAA1u2cfoiBhRBsUBB45RpsXmD0HACeTtUKdWANNs1t7eOTeD1w-YN+oypqG1tNrR0AInCeiMXCiC2aiDvvJIgAQgCiADEAPJkf5RDCINjAGKQgDkXSQDF6fThplOF3Umm0SAsbhsGgaPiwUAwIDQQyCIWU0ViQyAA)

Reply:
Good call out, I'll make this change now!